posthog/temporal/batch_exports/s3_batch_export.py

```diff
@@ -970,23 +982,456 @@ async def insert_into_s3_activity_from_stage(inputs: S3InsertInputs) -> RecordsC
             [field.with_nullable(True) for field in record_batch_schema]
         )
 
-        consumer = S3Consumer(
-            heartbeater=heartbeater,
-            heartbeat_details=details,
-            data_interval_end=data_interval_end,
+        consumer = ConcurrentS3Consumer(
             data_interval_start=data_interval_start,
-            writer_format=WriterFormat.from_str(inputs.file_format, "S3"),
+            data_interval_end=data_interval_end,
             s3_inputs=inputs,
+            part_size=settings.BATCH_EXPORT_S3_UPLOAD_CHUNK_SIZE_BYTES,
+            max_concurrent_uploads=settings.BATCH_EXPORT_S3_MAX_CONCURRENT_UPLOADS,
         )
-        _ = await run_consumer(
-            consumer=consumer,
+
+        records_completed = await run_consumer_from_stage(
             queue=queue,
+            consumer=consumer,
             producer_task=producer_task,
             schema=record_batch_schema,
-            max_bytes=settings.BATCH_EXPORT_S3_UPLOAD_CHUNK_SIZE_BYTES,
+            file_format=inputs.file_format,
+            compression=inputs.compression,
             include_inserted_at=True,
-            writer_file_kwargs={"compression": inputs.compression},
             max_file_size_bytes=inputs.max_file_size_mb * 1024 * 1024 if inputs.max_file_size_mb else 0,
+            json_columns=("properties", "person_properties", "set", "set_once"),
         )
 
-        return details.records_completed
+        return records_completed
+
+
+class ConcurrentS3Consumer(ConsumerFromStage):
+    """A consumer that uploads chunks of data to S3 concurrently.
+
+    It uses a memory buffer to store the data and upload it in parts. It uses 2 semaphores to limit the number of
+    concurrent uploads and the memory buffer.
+    """
+
+    def __init__(
+        self,
+        data_interval_start: dt.datetime | str | None,
+        data_interval_end: dt.datetime | str,
+        s3_inputs: S3InsertInputs,
+        part_size: int = 50 * 1024 * 1024,  # 50MB parts
+        max_concurrent_uploads: int = 5,
+        max_memory_buffer: int = 250 * 1024 * 1024,  # 250MB total buffer
```

Comment:
Shouldn't `max_memory_buffer` be `part_size * max_concurrent_uploads`?

Reply:
have removed this now